spring-rabbit/src/main/java/org/springframework/amqp/rabbit/config/BaseRabbitListenerContainerFactory.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.amqp.rabbit.config;
+
+import java.util.Arrays;
+
+import org.springframework.amqp.core.MessagePostProcessor;
+import org.springframework.amqp.rabbit.listener.AbstractMessageListenerContainer;
+import org.springframework.amqp.rabbit.listener.MessageListenerContainer;
+import org.springframework.amqp.rabbit.listener.RabbitListenerContainerFactory;
+import org.springframework.amqp.rabbit.listener.RabbitListenerEndpoint;
+import org.springframework.amqp.rabbit.listener.adapter.AbstractAdaptableMessageListener;
+import org.springframework.amqp.utils.JavaUtils;
+import org.springframework.lang.Nullable;
+import org.springframework.retry.RecoveryCallback;
+import org.springframework.retry.support.RetryTemplate;
+import org.springframework.util.Assert;
+
+/**
+ * Base abstract class for listener container factories.
+ *
+ * @param <C> the container type that the factory creates.
+ *
+ * @author Gary Russell
+ * @since 2.4
+ *
+ */
+public abstract class BaseRabbitListenerContainerFactory<C extends MessageListenerContainer>
+	implements RabbitListenerContainerFactory<C> {
+
+	protected Boolean defaultRequeueRejected;
+
+	private MessagePostProcessor[] beforeSendReplyPostProcessors;
+
+	private RetryTemplate retryTemplate;
+
+	private RecoveryCallback<?> recoveryCallback;
+
+	@Override
+	public C createListenerContainer(RabbitListenerEndpoint endpoint) {
+		return null;
```

Comment:
Why this is not `abstract` any more at this level?
Shouldn't every single implementor provide its specific implementation anyway?